app/routes.py

```diff
@@ -1 +1,239 @@
+
+# error handling
+def validate_model(cls, model_id):
+    try:
+        model_id = int(model_id)
+    except:
+        abort(make_response({"message":f"task {model_id} invalid"}, 400))
+
+    model = cls.query.get(model_id)
+
+    if not model:
+        abort(make_response({"details": f"{cls.__name__} id {model_id} not found"}, 404))
+
+    return model
+
+
+# ============================= CRUD ROUTES  =============================
+# CREATE TASK- POST request to /tasks
+@tasks_bp.route("", methods=["POST"])
+def create_task():
+    request_body = request.get_json()
+    
+    if "title" not in request_body or "description" not in request_body:
+        return {"details": "Invalid data"}, 400
+    
+    new_task = Task.from_dict(request_body)
+
+    db.session.add(new_task)
+    db.session.commit()
+
+    return new_task.to_dic(), 201
+    
+
+# READ ALL TASKS- GET request to /tasks
+@tasks_bp.route("", methods=["GET"])
+def read_all_tasks():
+    sorted_by_query = request.args.get("sort")
+    title_query = request.args.get("title")
+    
+    if sorted_by_query== "asc":
+        tasks = Task.query.order_by(Task.title).all()
+    elif sorted_by_query == "desc":
+        tasks = Task.query.order_by(Task.title.desc()).all()
+    elif sorted_by_query == "id":
+        tasks = Task.query.order_by(Task.task_id).all()
+    elif title_query:
+        tasks = Task.query.filter_by(title=title_query).all()
+    
+    if not sorted_by_query:  
+        tasks= Task.query.all()
+    
+    task_response = [task.to_dic()["task"] for task in tasks]
+    return jsonify(task_response), 200
+
+
+# READ ONE TASK- GET request to /tasks/<task_id>
+@tasks_bp.route("/<task_id>", methods=["GET"])
+def read_one_task(task_id):
+    task = validate_model(Task, task_id)
+    return task.to_dic(), 200
+
+
+# UPDATE TASK- PUT request to /tasks/<task_id>
+@tasks_bp.route("/<task_id>", methods=["PUT"])
+def update_task(task_id):
+    task = validate_model(Task, task_id)
+    request_body = request.get_json()
+    
+    task.title = request_body["title"]
+    task.description = request_body["description"]
+    
+    db.session.commit()
+    
+    return task.to_dic(), 200
+    
+    
+# mark complete on a complete task
+# UPDATE TASK- PATCH request to /tasks/<task_id>/mark_complete
+@tasks_bp.route("/<task_id>/mark_complete", methods=["PATCH"])
+def mark_complete(task_id):
+    task = validate_model(Task,task_id)
+    
+    task.completed_at=datetime.utcnow()
+    
+    db.session.commit()
+    
+    # ID of the channel you want to send the message to
+    channel_id = "C057AARLUNR"
+    
+    slack_url = "https://slack.com/api/chat.postMessage"
+    
+    slack_data = {
+        "channel": channel_id,
+        "text": f"Someone completed the {task.title} task!"
+    }
+    
+    headers= {
+        "Authorization": f"Bearer {os.environ.get('SLACK_KEY')}"
+    }
+    
+    slack_request = requests.post(slack_url, headers=headers, json=slack_data)
```

Comment:
Nice work on having your Slack post sent after the logic of marking a task complete! We don't want to send out any false positive alerts just in case our logic fails during the update!